hera_pspec/grouping.py

```diff
@@ -363,29 +365,33 @@ def average_spectra(uvp_in, blpair_groups=None, time_avg=False,
                     
                     # Take time average if desired
                     if time_avg:
+                        wsum = np.sum(w, axis=0).clip(1e-40, np.inf)
                         data = (np.sum(data * w, axis=0) \
-                            / np.sum(w, axis=0).clip(1e-40, np.inf))[None]
+                                / wsum)[None]
                         wgts = (np.sum(wgts * w[:, :1, None], axis=0) \
-                            / np.sum(w, axis=0).clip(1e-40, np.inf)[:1, None])[None]
+                                / wsum[:1, None])[None]
                         # wgts has a shape of (Ntimes, Nfreqs, 2), while 
                         # w has a shape of (Ntimes, Ndlys) or (Ntimes, 1)
                         # To handle with the case  when Nfreqs != Ntimes,
                         # we choose to multiply wgts with w[:,:1,None]. 
                         ints = (np.sum(ints * w, axis=0) \
-                             / np.sum(w, axis=0).clip(1e-40, np.inf))[None]
+                                / wsum)[None]
                         nsmp = np.sum(nsmp, axis=0)[None]
                         if store_window:
                             window_function = (np.sum(window_function * w[:, :, None], axis=0) \
-                                            / (np.sum(w,axis=0).clip(1e-40, np.inf))[:, None])[None]
+                                               / (wsum)[:, None])[None]
                         if store_cov:
                             cov = (np.sum(cov * w[:, :, None] * w[:, None, :], axis=0) \
-                                / (np.sum(w, axis=0).clip(1e-40, np.inf))[:, None] / (np.sum(w, axis=0).clip(1e-40, np.inf))[None, :])[None]
+                                   / wsum[:, None] / wsum[None, :])[None]
                         for stat in stat_l:
-                            # clip errws because it may be inf, and inf * 0 is nan
-                            errws[stat] = (np.sum(errws[stat].clip(0, 1e40) * w**2, axis=0) \
-                                        / (np.sum(w, axis=0).clip(1e-40, np.inf))**2)[None]
+                            # clip errws to eliminate nan: inf * 0 yields nans
+                            weighted_errws = errws[stat].clip(0, 1e40) * w**2
+                            errws[stat] = (np.sum(weighted_errws, axis=0) \
+                                           / wsum**2)[None]
+                            # set near-zero errws to inf, as they should be
+                            errws[stat][np.isclose(errws[stat], 0)] = np.inf
```

Comment:
If we have zero input for errws[stat] , then weighted_errws = errws[stat] * w**2 also goes to zero, but we could not set it to be np.inf here. Maybe we should just apply a cut (1e-40, 1e40) on both the lower end and higher end of weights w, in this way 1e-40*np.inf = inf and 1e40*0=0 leads to no bugs too. Since we have cuts on w on both ends, we don't need to apply cuts on other attributes wsum and weighted_errws here.

Reply:
Yeah so what you suggest is effectively what we do in computing the weights abovev: we clip the errws at (0, 1e-40). However, the errorbars are intrinsically allowed to be inf (e.g. for data w/ zero integration), so when we exit the errorbar calculation (as done here) we are setting errorbars that should be inf to inf (ones with zero integration), which is needed because with zero weights but clipped errws, the errorbars would otherwise be 0, which is incorrect.

Reply:
Yes, I mean there are two cases lead to a weighted errws[stat] to be 0, while the code  errws[stat][np.isclose(errws[stat], 0)] = np.inf seems to set both of them back to np.inf. Imagine errws[stat] to be 0 and np.inf, the clipped weights w is 1e40 and 0; thus weighted_errws = errws[stat].clip(0, 1e40) \* w\^2 is 0\*1e40=0 and 1e40\*0 = 0; wsum is 1e40 and 1e-40; errws[stat] = (np.sum(weighted_errws, axis=0) \
                                            / wsum^2) is 0/1e80 and 0/1e-80 = 0.

Reply:
Ah yes, but we would never expect `errws` to be 0 intrinsically. This is an unachievable scenario (t_int=>infinity) that is also somewhat pathological, as errws that are zero will lead to an inf weight for that data point, which mean the final average is informed solely from that data point. So I think its not an issue..